### PR TITLE
dev_guidelines: initiate a paragraph about unit-tests

### PR DESCRIPTION
**Pull request by goneri**
Add some basic information regarding the unit-testing and initiate a list
guidelines. More will come later.

### CODE REVIEW

File: docs/docsite/rst/dev_guidelines.rst [hunk 1]
```diff
@@ -1048,3 +1048,41 @@ Where one of these reasons apply you should open a pull request proposing the mi
...
 necessary policies should be available so that the tests can be manually run by
 someone performing a PR review or writing a patch.
+
+Unit-tests for AWS plugins
+==========================
+
+Why do we need unit-tests when we've got functional tests
+---------------------------------------------------------
+
+Unit-tests are much faster and more suitable to test corner cases. They are also don't depend on a third party service
```
Comment (tremble) [inline docs/docsite/rst/dev_guidelines.rst:1058]:
```suggestion
Unit-tests are much faster and more suitable to test corner cases. They also don't depend on a third party service
```

File: docs/docsite/rst/dev_guidelines.rst [hunk 2]
```diff
@@ -1048,3 +1048,57 @@ Where one of these reasons apply you should open a pull request proposing the mi
...
+and thus, a failure is less likley to be a false positive.
+
+
+How to keep my code simple?
+---------------------------
+
+Ideally, you should break up your code in tiny functions. Each function should have a limited number of parameters
+and a low amount of cross dependencies with the rest of the code (low coupling):
+
+- Don't pass a large data structure to a function if it only use one field. This clarify the inputs of your
```
Comment (jillr) [inline docs/docsite/rst/dev_guidelines.rst:1068]:
```suggestion
- Don't pass a large data structure to a function if it only use one field. This clarifies the inputs of your
```

File: docs/docsite/rst/dev_guidelines.rst [hunk 3]
```diff
@@ -1048,3 +1048,57 @@ Where one of these reasons apply you should open a pull request proposing the mi
...
+  function (the contract) and also reduce the risk of an unexpected transformation of the data structure
+  from within the function.
+- The boto client object is complex and can be source of unwanted side-effect. It's better to isolate the calls
+  in dedicated functions. These functions will have their own unit-tests.
+- Don't pass the ``module`` object when you only need the read a couple of parameters from ``module.params``.
+  Pass the parameter directly to your function. By doing so, you're explicite about what the function inputs
```
Comment (jillr) [inline docs/docsite/rst/dev_guidelines.rst:1074]:
```suggestion
  Pass the parameter directly to your function. By doing so, you're explicit about the function's inputs
```

File: docs/docsite/rst/dev_guidelines.rst [hunk 4]
```diff
@@ -1048,3 +1048,57 @@ Where one of these reasons apply you should open a pull request proposing the mi
...
+  (the contract) and you reduce potential side-effect.
+
+Unit-tests guidlines
+--------------------
+
+Ideally, all the ``module_utils`` should be covered by unit-tests. However we acknoledge that writing unit-tests may
```
Comment (jillr) [inline docs/docsite/rst/dev_guidelines.rst:1080]:
```suggestion
Ideally, all the ``module_utils`` should be covered by unit-tests. However we acknowledge that writing unit-tests may
```

File: docs/docsite/rst/dev_guidelines.rst [hunk 5]
```diff
@@ -1048,3 +1048,57 @@ Where one of these reasons apply you should open a pull request proposing the mi
...
+be challenging and we also accept contribution with no coverage. The test coverage of the other type of plugins is recommand.
```
Comment (jillr) [inline docs/docsite/rst/dev_guidelines.rst:1081]:
```suggestion
be challenging and we also accept contribution with no coverage. Test coverage of the other type of plugins is recommended.
```

File: docs/docsite/rst/dev_guidelines.rst [hunk 6]
```diff
@@ -1048,3 +1048,57 @@ Where one of these reasons apply you should open a pull request proposing the mi
...
+
+- Our tests are run with ``pytest`` and we use the features it provides such as Fixtures, Parametrization.
+- The use of ``unittest.TestCase`` is discouraged for the sack of consistency and simplicity.
```
Comment (jillr) [inline docs/docsite/rst/dev_guidelines.rst:1084]:
```suggestion
- The use of ``unittest.TestCase`` is discouraged for the sake of consistency and simplicity.
```

File: docs/docsite/rst/dev_guidelines.rst [hunk 7]
```diff
@@ -1048,3 +1048,57 @@ Where one of these reasons apply you should open a pull request proposing the mi
...
+- Unit-tests should run fine without any network connection.
+- It's not necessary to mock all the boto3/botocore calls (``get_paginator()``, ``paginate()``, etc). It's often better to just set-up a function that wrap these calls and mock the result.
+- Simplicity prevails. Tests should be short and cover a limited set of features.
+
+Pytest is well documented and you will find some example in its `ow-to guides <https://docs.pytest.org/en/latest/how-to/index.html>`_
```
Comment (jillr) [inline docs/docsite/rst/dev_guidelines.rst:1089]:
```suggestion
Pytest is well documented and you will find some example in its `how-to guides <https://docs.pytest.org/en/latest/how-to/index.html>`_
```

**Review by jillr — COMMENTED** [review]
Some copy-editing nits - thanks Goneri!

File: docs/docsite/rst/dev_guidelines.rst [hunk 8]
```diff
@@ -1048,3 +1048,57 @@ Where one of these reasons apply you should open a pull request proposing the mi
...
+  Pass the parameter directly to your function. By doing so, you're explicite about what the function inputs
+  (the contract) and you reduce potential side-effect.
+
+Unit-tests guidlines
+--------------------
+
+Ideally, all the ``module_utils`` should be covered by unit-tests. However we acknoledge that writing unit-tests may
+be challenging and we also accept contribution with no coverage. The test coverage of the other type of plugins is recommand.
+
+- Our tests are run with ``pytest`` and we use the features it provides such as Fixtures, Parametrization.
```
Comment (alinabuzachis) [inline docs/docsite/rst/dev_guidelines.rst:1083]: Wouldn't make sense to provide a generic example for Fixtures and Parametrization?

File: docs/docsite/rst/dev_guidelines.rst (hunk 6 above)
Comment (goneri) [reply docs/docsite/rst/dev_guidelines.rst:1084]: whoa, that was a nice one :-).

File: docs/docsite/rst/dev_guidelines.rst (hunk 8 above)
Comment (goneri) [reply docs/docsite/rst/dev_guidelines.rst:1083]: Pytest documentation is pretty good and cover these features. But we can add some links pointing on was we identify as our best example of well written unit-tests.
This being said, I would suggest to do this in another patch.

File: docs/docsite/rst/dev_guidelines.rst [hunk 9]
```diff
@@ -1048,3 +1048,57 @@ Where one of these reasons apply you should open a pull request proposing the mi
...
+- Don't pass the ``module`` object when you only need the read a couple of parameters from ``module.params``.
+  Pass the parameter directly to your function. By doing so, you're explicite about what the function inputs
+  (the contract) and you reduce potential side-effect.
+
+Unit-tests guidlines
+--------------------
+
+Ideally, all the ``module_utils`` should be covered by unit-tests. However we acknoledge that writing unit-tests may
+be challenging and we also accept contribution with no coverage. The test coverage of the other type of plugins is recommand.
+
```
[the remark this answers is not in the document]
Comment (goneri) [reply docs/docsite/rst/dev_guidelines.rst:1082]: Hum, indeed. I've rephrased a bit the the sentence to clarify we only speak about unit-test.

File: docs/docsite/rst/dev_guidelines.rst (hunk 8 above)
Comment (tremble) [reply docs/docsite/rst/dev_guidelines.rst:1083]: The pytest docs are very good once you understand what you're looking at, but I found some of the examples a little too abstracted from reality to understand at first especially as it applies to our specific modules and module_utils.

File: docs/docsite/rst/dev_guidelines.rst [hunk 10]
```diff
@@ -1048,3 +1048,57 @@ Where one of these reasons apply you should open a pull request proposing the mi
...
+- Don't pass a large data structure to a function if it only use one field. This clarifies the inputs of your
+  function (the contract) and also reduce the risk of an unexpected transformation of the data structure
+  from within the function.
+- The boto client object is complex and can be source of unwanted side-effect. It's better to isolate the calls
+  in dedicated functions. These functions will have their own unit-tests.
+- Don't pass the ``module`` object when you only need the read a couple of parameters from ``module.params``.
+  Pass the parameter directly to your function. By doing so, you're explicit about the function's inputs
+  (the contract) and you reduce potential side-effect.
+
+Unit-tests guidlines
```
Comment (tremble) [inline docs/docsite/rst/dev_guidelines.rst:1077]:
```suggestion
Unit-tests guidelines
```

File: docs/docsite/rst/dev_guidelines.rst [hunk 11]
```diff
@@ -1048,3 +1048,57 @@ Where one of these reasons apply you should open a pull request proposing the mi
...
+
+
+How to keep my code simple?
+---------------------------
+
+Ideally, you should break up your code in tiny functions. Each function should have a limited number of parameters
+and a low amount of cross dependencies with the rest of the code (low coupling):
+
+- Don't pass a large data structure to a function if it only use one field. This clarifies the inputs of your
+  function (the contract) and also reduce the risk of an unexpected transformation of the data structure
```
Comment (tremble) [inline docs/docsite/rst/dev_guidelines.rst:1069]:
```suggestion
  function (the contract) and also reduces the risk of an unexpected transformation of the data structure
```

**Review by tremble — APPROVED** [review]

File: docs/docsite/rst/dev_guidelines.rst [hunk 12]
```diff
@@ -1048,3 +1048,57 @@ Where one of these reasons apply you should open a pull request proposing the mi
...
+  function (the contract) and also reduces the risk of an unexpected transformation of the data structure
+  from within the function.
+- The boto client object is complex and can be source of unwanted side-effect. It's better to isolate the calls
+  in dedicated functions. These functions will have their own unit-tests.
+- Don't pass the ``module`` object when you only need the read a couple of parameters from ``module.params``.
+  Pass the parameter directly to your function. By doing so, you're explicit about the function's inputs
+  (the contract) and you reduce potential side-effect.
+
+Unit-tests guidelines
+--------------------
```
Comment (tremble) [inline docs/docsite/rst/dev_guidelines.rst:1078]:
```suggestion
---------------------
```

**Review by tremble — APPROVED** [review]

File: docs/docsite/rst/dev_guidelines.rst [hunk 13]
```diff
@@ -1048,3 +1048,57 @@ Where one of these reasons apply you should open a pull request proposing the mi
...
+and thus, a failure is less likley to be a false positive.
+
+
+How to keep my code simple?
+---------------------------
+
+Ideally, you should break up your code in tiny functions. Each function should have a limited number of parameters
+and a low amount of cross dependencies with the rest of the code (low coupling):
+
+- Don't pass a large data structure to a function if it only use one field. This clarifies the inputs of your
```
Comment (alinabuzachis) [inline docs/docsite/rst/dev_guidelines.rst:1068]:
```suggestion
- Don't pass a large data structure to a function if it only uses one field. This clarifies the inputs of your
```

File: docs/docsite/rst/dev_guidelines.rst [hunk 14]
```diff
@@ -1048,3 +1048,57 @@ Where one of these reasons apply you should open a pull request proposing the mi
...
 someone performing a PR review or writing a patch.
+
+Unit-tests for AWS plugins
+==========================
+
+Why do we need unit-tests when we've got functional tests
+---------------------------------------------------------
+
+Unit-tests are much faster and more suitable to test corner cases. They also don't depend on a third party service
+and thus, a failure is less likley to be a false positive.
```
Comment (alinabuzachis) [inline docs/docsite/rst/dev_guidelines.rst:1059]:
```suggestion
and thus, a failure is less likely to be a false positive.
```

**Review by alinabuzachis — APPROVED** [review]
Two last typos, otherwise LGTM! Thank you @goneri
